File: Baek-joon/11053/11053.py
import sys
input = sys.stdin.readline

# 가장 긴 증가하는 수열

# 찾고자 하는 값의 이상인 곳이 처음으로 나타나는 곳을 반환한다.
def lower_bound(arr, start, end,  check) :
    while end - start > 0 :
        mid = int((start + end) / 2)
        if arr[mid] < check :
            start = mid + 1
        else :
            end = mid
    return end+1

if __name__ == "__main__":
    n = int(input().strip())
    arr = list(map(int,input().strip().split()))
    lis = [-float('INF')]
    i=0
    # 일단 가장 긴 증가하는 순열을 찾기 위해 배열의 시작을 넣어준다.
    lis[i] = arr[i]
    i = i+1

    while i<n :
        # 만약 현재의 LIS 배열의 마지막 숫자(현재 LIS에서 가장 큰 수)가 더 작다면 그냥 뒤에 붙힌다.
        if lis[-1]<arr[i] :
            lis.append(arr[i])
        else :
            # 아니라면 그냥 지금 수가 현재의 수보다 가장 처음으로 커지는 곳에 대신 집어 넣는다.
            ans = lower_bound(lis, 0,len(lis),arr[i])
            lis[ans - 1] = arr[i]
        i = i + 1

    print(len(lis))

    # 주의점 LIS의 길이를 알 수 있을뿐 LIS 배열을 알고 싶을땐 이 방법이 아니다.

File: Algorithm_problem_solving/Baek-joon/11053/11053.kt
/**
 * 가장 긴 부분 순열(lis)를 계산하는 클래스
 */
class LongestIncreasingSubsequence {
    /**
     * 가장 긴 부분 순열의 길이를 구한다.
     */
    fun lisLength(arr: LongArray): Int {
        val lis = LongArray(arr.size) { 0 }
        var lastIndex = -1
        arr.forEach {
            if (lastIndex == -1) {
                // 배열이 비어있다면 넣는다.
                lastIndex += 1
                lis[lastIndex] = it
            } else if (it > lis[lastIndex]) {
                // 마지막 값보다 현재 값이 크다면 가장 뒤에 붙힌다.
                lastIndex += 1
                lis[lastIndex] = it
            } else {
                lis[lowerBound(lis, it, 0, lastIndex)] = it
            }
        }
        return lastIndex + 1
    }

    /**
     * @param arr 정렬된 리스트
     * @param find 찾으려는 값
     * @param start 찾으려는 범위의 시작 index
     * @param end 찾으려는 범위의 마지막 index
     * @return 리스트에서 find 보다 커지는 값이 나오는 첫번째 위치
     */
    private fun lowerBound(
        arr: LongArray,
        find: Long,
        start: Int,
        end: Int,
    ): Int {
        var s = start
        var e = end
        while (s < e) {
            val mid = (s + e) / 2
            if (arr[mid] < find) {
                s = mid + 1
            } else {
                e = mid
            }
        }
        return e
    }
}


fun main() {
    val n = readLine()?.toInt() ?: throw NumberFormatException()
    val arr = readLine()?.split(" ")?.map { it.toLong() }?.toLongArray() ?: throw NumberFormatException()

    println(
        LongestIncreasingSubsequence().lisLength(arr)
    )
}